arai60/54-60_others/59_8_String to Integer (atoi)/level_2.py

```diff
@@ -0,0 +1,84 @@
+# 各処理を関数化
+# オーバーフローの判定をint範囲内に収めるように変更
+class Solution:
+    def myAtoi(self, s: str) -> int:
+        INT_MAX = 2**31 - 1
+        INT_MIN = -(2**31)
+
+        def skip_space(index):
+            while index < len(s) and s[index].isspace():
+                index += 1
+            return index
+
+        def parse_sign(index):
+            sign = 1
+            if index < len(s) and (s[index] == "-" or s[index] == "+"):
+                if s[index] == "-":
+                    sign = -1
+                return index + 1, sign
+            return index, sign
+
+        def is_overflow(current_value, sign, next_digit):
+            if sign == 1:
+                if current_value > INT_MAX // 10:
+                    return True
+                if current_value == INT_MAX // 10 and next_digit >= 7:
```

Comment:
```suggestion
                if current_value == INT_MAX // 10 and next_digit >= INT_MAX % 10:
```

せっかくINT_MAXを定義しているので1の位の値でオーバーフローするかどうか判定する部分もINT_MAXを使って書きたい気持ちがあります。
64ビットについても対応が必要などの改修が必要になった際の保守性が良くなるというのが理由でしょうか。



sign = -1のときも同様です。こちらは Pythonの`%`演算子の挙動も合わせて確認しておくと良いと思います。

Reply:
Pythonだとこの部分が確認して置いたほうが良いかなと思います。

> The modulo operator always yields a result with the same sign as its second operand 

`%`のマイナスで割る際の結果は言語ごとに違います。`/`で切り捨てる際に0方向か負の無限大方向に切り捨てるかの話も同様ですね。

Reply:
Haskellなどmoduloとreminderを明確に区別する言語もあるみたいです。

https://zvon.org/other/haskell/Outputprelude/mod_f.html
https://zvon.org/other/haskell/Outputprelude/rem_f.html